Applications/ApplicationsLib/ProjectData.cpp

```diff
@@ -352,3 +355,41 @@ void ProjectData::parseNonlinearSolvers(BaseLib::ConfigTree const& config)
 		    "The nonlinear solver name is not unique");
 	}
 }
+
+static std::unique_ptr<MathLib::PiecewiseLinearInterpolation>
+createPiecewiseLinearInterpolation(BaseLib::ConfigTree const& config)
+{
+	auto const coords = config.getConfParam<std::vector<double>>("coords");
+	auto const values = config.getConfParam<std::vector<double>>("values");
+	if (coords.empty() || values.empty())
+	{
+		ERR("The given co-ordinates or values vector is empty.");
+		std::abort();
+	}
+	if (coords.size() != values.size())
+	{
+		ERR("The given co-ordinates and values vector sizes are different.");
+		std::abort();
+	}
+
+	return std::unique_ptr<MathLib::PiecewiseLinearInterpolation>{
+	    new MathLib::PiecewiseLinearInterpolation{coords, values}};
```

Comment:
> cf. to "Effective modern c++" Item 41: "Consider pass by value for copyable parameters that are cheap to move and always copied."

But we do not always copy the vectors. In fact, they are only copied in the unit tests, aren't they?

My opinion:
1. IMHO the best option
2. Problematic. Maybe then different calls of `std::vector`'s ctor other than copy/move could be generated implicitly, e.g., `vector(std::initializer_list)`, though I'm not sure about it. Anyway in this case the avoided code duplication is neglectable. Furthermore we don't do such things in any other place, do we?
3. There will always be a copy and we rely on the compilers optimizer to perform the copy elision.
4. There are have several other places where we've only implemented r-value reference arguments, so it wouldn't hurt (unless you can give an example that hurts :smile:).

Anyway, that's only a minor issue. Don't waste time on it. If you can't decide on a specific solution, please just leave a TODO comment in the code, so we can find the spot later on. I'd vote for option 1 or 4, choose either one of them, if you like.